modules/axp/Animal/lib/Animal.pm

```diff
@@ -22,76 +26,55 @@ our $VERSION = '0.01';
 
 Don't use this class directly; instead instantiate subclasses of it.
 
-=head1 EXPORT
-
 =head1 SUBROUTINES/METHODS
 
-=head2 named
-
-Construct a new Animal, with the given name
-
-=cut
-
-sub named {
-	ref ( my $class = shift ) and croak "Static constructor used as instance call";
-	my $name = shift or croak "Need to provide a name";
-	my $self = { Name => $name, Colour => $class->default_colour };
-	return bless $self, $class;
-}
-
 =head2 name
 
 Get or set the Animal's name
 
 =cut
 
-sub name {
-	ref ( my $self = shift ) or croak "Instance variable needed";
-	if ( @_ ) {
-		$self->{ Name } = shift;
-		return $self;
-	}
-	else {
-		return $self->{ Name };
-	}
-}
+has 'name' => ( is => 'rw', required => 1 );
+
+before 'name' => $LivingCreature::__instance_check;
 
 =head2 colour
 
 Get or set the Animal's colour
 
 =cut
 
-sub colour {
+has 'colour' => ( is => 'rw', default => sub {
+	return ref ( shift )->default_colour;
+});
+
+around 'colour' => sub {
+	my $orig = shift;
 	my $either = shift;
 	
-	if ( @_ ) {
-		ref ( $either ) or croak "Instance variable needed";
-		$either->{ Colour } = shift;
-		return $either;
-	}
-	else {
-		return ref ( $either ) ? $either->{ Colour } : $either->default_colour();
+	if ( !ref ( $either ) ) {
+		croak "Instance variable needed" if shift;
+		return $either->default_colour;
 	}
-}
+	
+	return $either->$orig( @_ );
```

Comment:
OK, so this is the ugliest part of the solution. If we call `colour` on a class rather than an instance, we want to short-circuit Moose's own method and instead defer to `default_colour`. This requires an `around` hook.